mlx/transforms.cpp

```diff
@@ -216,13 +218,43 @@ void eval(const std::vector<array>& outputs) {
   auto synchronizer =
       array({}, bool_, std::make_unique<Synchronizer>(stream), outputs);
 
-  recurse = [&](const array& a) {
+  compute_depths = [&](const array& a) {
+    auto id = a.id();
+    if (auto it = depth_map.find(id); it != depth_map.end()) {
+      return it->second;
+    }
+    int depth = 0;
+    for (auto& in : a.inputs()) {
+      depth = std::max(depth, compute_depths(in));
+    }
+    depth++;
+    depth_map[id] = depth;
+    for (auto& s : a.siblings()) {
+      depth_map[s.id()] = depth;
+    }
+    return depth;
+  };
+
+  recurse = [&](const array& a, bool largest_branch_first) {
     auto id = a.id();
     if (cache.find(id) != cache.end()) {
       return;
     }
-    for (auto in : a.inputs()) {
-      recurse(in);
+    std::vector<int> recursion_order(a.inputs().size());
+    std::iota(recursion_order.begin(), recursion_order.end(), 0);
+    std::sort(
+        recursion_order.begin(),
+        recursion_order.end(),
+        [&a, &depth_map, largest_branch_first](int left, int right) {
+          int depth_left = depth_map[a.inputs()[left].id()];
+          int depth_right = depth_map[a.inputs()[right].id()];
+          bool large_left = depth_left >= depth_right;
+          return largest_branch_first == large_left;
```

Comment:
Nit: the naming here confused me a little, makes it sound like a binary tree. Maybe change `left`, `right` to simply `i` and `j` (or something like that)

Reply:
Changed to `i`, `j`. Honestly, I think `left`, `right` was less confusing. Left as in smaller index in the array. Return true if it shouldn't be swapped.

The whole sort API is kinda bad tbh. For instance the above check can lead to crashes because it has larger or equal instead of larger. It is now fixed as well.

Reply:
> Changed to i, j. Honestly, I think left, right was less confusing. Left as in smaller index in the array

I see.. maybe I misunderstand the invariants of `std::sort`, but I thought `i` and `j` were just arbitrary indices, not not necessarily true that `i  < j`.

Reply:
Looking back into the docs, it actually is neither. The docs say `comp(*(it + n), *it)` should return `false` for any valid `it` and `n` after sort is done. So I guess if `comp(left, right)` should return `true` if we need to swap.